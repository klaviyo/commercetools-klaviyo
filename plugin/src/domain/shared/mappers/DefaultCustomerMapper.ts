import { Address, Customer } from '@commercetools/platform-sdk';
import { CustomerMapper } from './CustomerMapper';
import { mapAllowedProperties } from '../../../utils/property-mapper';

const E_164_REGEX = /^\+[1-9]\d{10,14}$/;

export class DefaultCustomerMapper implements CustomerMapper {
    public mapCtCustomerToKlaviyoProfile(customer: Customer, klaviyoProfileId?: string): ProfileRequest {
        const address = this.getCTCustomerAddressForKlaviyo(customer);
        const {
            email,
            firstName: first_name,
            lastName: last_name,
            title,
            companyName: organization,
            id: external_id,
            custom,
        } = customer;
        const props = mapAllowedProperties('customer.customFields', { ...(custom?.fields || {}) });

        return {
            data: {
                type: 'profile',
                id: klaviyoProfileId,
                attributes: {
                    email,
                    external_id,
                    first_name,
                    last_name,
                    title,
                    phone_number: address?.mobile || address?.phone,
                    organization,
                    location: this.mapCTAddressToKlaviyoLocation(address),
                    properties: Object.keys(props).length ? { ...props } : undefined,
                },
            },
        };
    }

    public mapCTAddressToKlaviyoLocation(address?: Address, useSpecialPrefix = false): KlaviyoLocation | null {
        const specialPrefix = useSpecialPrefix ? '$' : '';
        return address
            ? {
                  [`${specialPrefix}address1`]: this.getAddressLine1(address),
                  [`${specialPrefix}address2`]: this.getAddressLine2(address),
                  [`${specialPrefix}city`]: address.city,
                  [`${specialPrefix}country`]: address.country,
                  [`${specialPrefix}region`]: address.region,
                  [`${specialPrefix}zip`]: address.postalCode,
              }
            : null;
    }

    public getAddressLine1(address?: Address): string | null{
        return address
            ? [address.apartment, address.building, address.streetNumber, address.streetName]
                  .filter((element) => element)
                  .join(', ')
            : null;
    }

    public getAddressLine2(address?: Address): string | null {
        return address
            ? [address.additionalStreetInfo, address.additionalAddressInfo].filter((element) => element).join(', ')
            : null;
    }

    // No longer in use, validation only handled by Klaviyo.
    // Phone number removed in worse case scenario.
    public getPhoneNumber(address?: Address): string | null {
        const phoneNumber = address?.mobile || address?.phone;
        if (phoneNumber) {
            const noSpacesAndDashPhoneNumber = phoneNumber.replace(/\s|-|\/|\)|\(|\*|[a-z]|[A-Z]+/g, '');
            return E_164_REGEX.test(noSpacesAndDashPhoneNumber) ? noSpacesAndDashPhoneNumber : null;
        }
        return null;
    }

    /**
     * A CT customer can have multiple addresses, klaviyo accepts only one address information in the location field.
     * This method selects one of the CT addresses to be used to populate the klaviyo location info for a klaviyo profile.
     * The address selection evaluates in order the following conditions and returns the first address found:
     * 1. Selects the customer defaultBillingAddress if set
     * 2. Selects the first available billing address if present
     * 3. Selects the first available address if present
     * @param customer
     */
    public getCTCustomerAddressForKlaviyo(customer: Customer): Address | undefined {
        if (!customer || !customer.addresses.length) {
            return;
        }
        if (customer.defaultBillingAddressId) {
            return customer.addresses?.find((address) => address.id === customer.defaultBillingAddressId);
        }
        if (customer?.billingAddressIds?.length) {
            return customer.addresses?.find((address) => address.id === customer?.billingAddressIds?.find(() => true));
        }
        return customer.addresses?.find(() => true);
    }
}
