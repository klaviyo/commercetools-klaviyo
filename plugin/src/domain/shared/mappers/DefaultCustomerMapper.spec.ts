import { DefaultCustomerMapper } from './DefaultCustomerMapper';
import { sampleCustomerApiResponse } from '../../../test/testData/ctCustomerApi'
import { expect as exp } from 'chai';

const customerMapper = new DefaultCustomerMapper();
describe('map CT customer to Klaviyo profile', () => {
    it('should map a commercetools customer to a klaviyo profile', () => {
        const klaviyoEvent = customerMapper.mapCtCustomerToKlaviyoProfile(sampleCustomerApiResponse);
        expect(klaviyoEvent).toMatchSnapshot();
    });

    it('should include/exclude/map custom fields and send them with the klaviyo profile', () => {
        const klaviyoEvent = customerMapper.mapCtCustomerToKlaviyoProfile({
            ...sampleCustomerApiResponse,
            custom: {
                type: {
                    typeId: 'type',
                    id: '123456',
                },
                fields: {
                    includedField: true,
                    ignoredField: 'test-ignored-field',
                },
            }
        });
        expect(klaviyoEvent).toMatchSnapshot();
    });

    it('should map a commercetools customer to a klaviyo profile with an existing profile id', () => {
        const klaviyoEvent = customerMapper.mapCtCustomerToKlaviyoProfile(sampleCustomerApiResponse, '123456');
        expect(klaviyoEvent).toMatchSnapshot();
    });
});

describe('validate phone number', () => {
    it('should return null when the phone number is an empty string', () => {
        const result = customerMapper.getPhoneNumber({ country: 'aCountry', mobile: '' });
        exp(result).to.be.null;
    });
    it.each`
        phoneNumber
        ${'+12/345(678)*900'}
        ${'+12()*- 345678900'}
        ${'abcde()*()+12 345 678 900'}
        ${'+1-2-3-4-5-6-7-8-9-0-0'}
    `('should return the phone number when it is valid and remove symbols and spaces', ({ phoneNumber }) => {
        const result = customerMapper.getPhoneNumber({ country: 'aCountry', mobile: phoneNumber });
        exp(result).to.be.eq('+12345678900');
    });
    it('should return null when the phone number is invalid', () => {
        const result = customerMapper.getPhoneNumber({ country: 'aCountry', mobile: '31 102455400' });
        exp(result).to.be.null;
    });
    it('should return the mobile number when both phone and mobile are present in the address', () => {
        const result = customerMapper.getPhoneNumber({ country: 'aCountry', mobile: '+31 102455400', phone: '+31 102455401' });
        exp(result).to.be.eq('+31102455400');
    });
    it('should return the phone number when the mobile number is missing in the address', () => {
        const result = customerMapper.getPhoneNumber({ country: 'aCountry', phone: '+31 102455401' });
        exp(result).to.be.eq('+31102455401');
    });
});
