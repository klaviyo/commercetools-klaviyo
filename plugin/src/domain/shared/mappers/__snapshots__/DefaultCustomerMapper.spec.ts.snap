// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`map CT customer to Klaviyo profile should include/exclude/map custom fields and send them with the klaviyo profile 1`] = `
{
  "data": {
    "attributes": {
      "email": "Alfred_Lebsack@hotmail.com",
      "external_id": "77d8e6f4-2dee-4ac9-bd21-4035b6ba76d8",
      "first_name": "Jailyn",
      "last_name": "Hilll",
      "location": {
        "address1": "12, First Street",
        "address2": "",
        "city": "Brookefort",
        "country": "NL",
        "region": undefined,
        "zip": "12345",
      },
      "organization": undefined,
      "phone_number": "+39 3234567892",
      "properties": {
        "includedFieldAfterMap": true,
      },
      "title": "Miss",
    },
    "id": undefined,
    "type": "profile",
  },
}
`;

exports[`map CT customer to Klaviyo profile should map a commercetools customer to a klaviyo profile 1`] = `
{
  "data": {
    "attributes": {
      "email": "Alfred_Lebsack@hotmail.com",
      "external_id": "77d8e6f4-2dee-4ac9-bd21-4035b6ba76d8",
      "first_name": "Jailyn",
      "last_name": "Hilll",
      "location": {
        "address1": "12, First Street",
        "address2": "",
        "city": "Brookefort",
        "country": "NL",
        "region": undefined,
        "zip": "12345",
      },
      "organization": undefined,
      "phone_number": "+39 3234567892",
      "properties": undefined,
      "title": "Miss",
    },
    "id": undefined,
    "type": "profile",
  },
}
`;

exports[`map CT customer to Klaviyo profile should map a commercetools customer to a klaviyo profile with an existing profile id 1`] = `
{
  "data": {
    "attributes": {
      "email": "Alfred_Lebsack@hotmail.com",
      "external_id": "77d8e6f4-2dee-4ac9-bd21-4035b6ba76d8",
      "first_name": "Jailyn",
      "last_name": "Hilll",
      "location": {
        "address1": "12, First Street",
        "address2": "",
        "city": "Brookefort",
        "country": "NL",
        "region": undefined,
        "zip": "12345",
      },
      "organization": undefined,
      "phone_number": "+39 3234567892",
      "properties": undefined,
      "title": "Miss",
    },
    "id": "123456",
    "type": "profile",
  },
}
`;
